source/common/router/router.cc

```diff
@@ -470,6 +476,40 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::HeaderMap& trailers) {
   return Http::FilterTrailersStatus::StopIteration;
 }
 
+Http::FilterMetadataStatus Filter::decodeMetadata(Http::MetadataMap& metadata_map) {
+  Http::MetadataMap metadata;
+  metadata.insert(metadata_map.begin(), metadata_map.end());
+  Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata);
```

Comment:
So just to sanity check, we don't need to do the copy for headers because the HCM permanently stores request headers, but we copy metadata because right now the HCM doesn't store metadata, yes?

If so this will get fixed when we handle metadata retries and I'm good with both not creating a copy utility and Ok with the temporary perf hit.  Might be worth a comment that the copy will go away as we continue with metadata functionality.

Reply:
That's correct! I changed this copy to move since router is the last filter and no one will refer to metadata_map afterwards. In the future, the move will happen in HCM.